Fc 469 initiate recruit from my preferences page

ben1stone-leftdfe opened the pull request:
First commit contains the bulk of the work and unit tests and the second contains the integration tests.

ben1stone-leftdfe committed:
Reorganising

ben1stone-leftdfe committed:
Added integration tests

leewadhams commented on src/web/EmployerFavouritesWeb/Application/Queries/EmployerHasLegalEntityQuery.cs:
```diff
@@ -0,0 +1,10 @@
+﻿
```
Blank line

leewadhams commented on src/web/EmployerFavouritesWeb/Application/Queries/EmployerHasLegalEntityQueryHandler.cs:
```diff
@@ -0,0 +1,42 @@
+﻿using System.Threading;
+using System.Threading.Tasks;
+using DfE.EmployerFavourites.Domain;
```
Unused using

leewadhams commented on src/web/EmployerFavouritesWeb/Application/Queries/EmployerHasLegalEntityQueryHandler.cs:
```diff
@@ -0,0 +1,42 @@
+        {
+            _accountApi = accountApiClient;
+            _logger = logger;
+        }
```
Space between methods

leewadhams commented on src/web/EmployerFavouritesWeb/Application/Queries/EmployerHasLegalEntityQueryHandler.cs:
```diff
@@ -0,0 +1,42 @@
+
+                return LegalEntities.Count > 0;
+            }
+            catch (System.Exception e)
```
Would normally have a `using System` statement at the top of the file.

leewadhams commented on src/web/EmployerFavouritesWeb/Configuration/ExternalLinks.cs:
```diff
@@ -7,5 +7,7 @@ public class ExternalLinks
         public string AccountsHomePage { get; set; }
         public string AccountsDashboardPage { get; set; }
         public string AccountsRegistrationPage { get; set; }
+        public string RecruitFromProgramme { get; set; }
```
I think maybe they could be named to make it clearer what they're for. Maybe `CreateVacancyForProgramme` & `CreateVacancyForProgrammeAndProvider`.

What do you think?

Are they 2 separate urls in any case? or is it a single url that has optional query string parameters?

leewadhams commented on src/web/EmployerFavouritesWeb/Controllers/ApprenticeshipsController.cs:
```diff
@@ -97,6 +104,11 @@ public async Task<IActionResult> TrainingProvider(string employerAccountId, stri
                 ApprenticeshipId = apprenticeshipId
             });
 
+            var employerHasLegalEntityResponse = await _mediator.Send(new EmployerHasLegalEntityQuery
```
I'm wondering if we should do this in the controller of whether it should be part of the other query.

ben1stone-leftdfe replied:
I've finished all the other changes. Just this one left to do. I did consider it but it would need to be implemented into two slightly different queries one for Favourites and then again for the Providers when you go into each favourite.

ben1stone-leftdfe replied:
@leewadhams see above.

leewadhams commented on src/web/EmployerFavouritesWeb/Models/ApprenticeshipFavouritesViewModel.cs:
```diff
@@ -6,5 +6,7 @@ public class ApprenticeshipFavouritesViewModel
     {
         public IReadOnlyList<ApprenticeshipFavouriteViewModel> Items { get; set; }
         public IEnumerable<char> EmployerAccountId { get; set; }
+        public string ProgrammeUrl { get; set; }
```
CreateVacancyUrl?

leewadhams commented on src/web/EmployerFavouritesWeb/Models/TrainingProvidersViewModel.cs:
```diff
@@ -6,5 +6,8 @@ public class TrainingProvidersViewModel
     {
         public string EmployerAccountId { get; set; }
         public IReadOnlyList<TrainingProviderViewModel> Items { get; set; }
+        public string ProgrammeAndProviderUrl { get; set; }
```
CreateVacancyUrl?

leewadhams commented on src/web/EmployerFavouritesWeb/Views/Apprenticeships/TrainingProvider.cshtml:
```diff
@@ -83,10 +83,14 @@
+                        <div class="govuk-grid-column-one-third">
+                            @if (Model.HasLegalEntity)
+                            {
+                                <a href="@string.Format(Model.ProgrammeAndProviderUrl,Model.EmployerAccountId,Model.Items[i].Ukprn,Model.ApprenticeshipId)" class="govuk-button app-task-list__training-provider-content recruit-button">Recruit apprentice</a>
```
I'd be tempted to have the generated url as a property on the view model.

leewadhams commented on src/web/IntegrationTests/Stubs/StubAccountApiClient.cs:
```diff
@@ -40,7 +40,28 @@ public Task<EmployerAgreementView> GetEmployerAgreement(string accountId, string
+                    new ResourceViewModel
+                    {
+                        Id = "654321",
+                        Href = "/api/accounts/XXX321/legalentities/654321"
```
Don't know if it's confusing having different account ids in each of the urls.

leewadhams commented on src/web/IntegrationTests/appSettings.json:
```diff
@@ -41,7 +41,10 @@
     "AccountsDashboardPage": "https://test-accounts-dashboard/accounts/{0}/teams",
-    "AccountsRegistrationPage": ""
+    "AccountsRegistrationPage": "",
+    "RecruitFromProgramme": "https://recruit.at-eas.apprenticeships.education.gov.uk/accounts/{0}/employer-create-vacancy?ProgrammeId={1}",
```
Yeah I think we can probably just have a single url

leewadhams reviewed (changes requested):
Looks good. A few tidy up things to be done/considered.

ben1stone-leftdfe committed:
Updates following code review.
- Spacing corrections
- Combines RecruitFromProgramme and RecruitFromProgrammeAndProvider into one CreateVacancyUrl
- Added a loop in the controller to generate each CreateVacancyUrl rather than building in view
- Corrected AccountIds in stub

ben1stone-leftdfe committed:
Moved EmployerHasLegalEntityQuery into ViewEmployerFavourites and ViewTrainingProviders

ben1stone-leftdfe committed:
Updated to show both a view training providers button and recruit link rather than just replacing the view training providers button

leewadhams commented on src/api/IntegrationTests/IntegrationTests.csproj:
```diff
@@ -1,18 +1,20 @@
     <IsPackable>false</IsPackable>
     <RootNamespace>DfE.EmployerFavourites.Api.IntegrationTests</RootNamespace>
     <AssemblyName>DfE.EmployerFavourites.IntegrationTests</AssemblyName>
+    <UserSecretsId>e4b53c92-d5f5-45a6-9912-353d3945751a</UserSecretsId>
```
Don't think this is needed for this project.

ben1stone-leftdfe committed:
Update following code review

sultansfa1 reviewed (approved):
This has been tested and ready for release
